src/Algebra/Graph/Undirected.hs

```diff
@@ -0,0 +1,926 @@
+-- @
+edge :: a -> a -> Graph a
+edge = coerce2 G.edge
+
+-- | /Overlay/ two graphs. This is a
+-- commutative, associative and idempotent operation with the identity 'empty'.
+-- Complexity: /O(1)/ time and memory, /O(s1 + s2)/ size.
+--
+-- @
+-- 'isEmpty'     (overlay x y) == 'isEmpty'   x   && 'isEmpty'   y
+-- 'hasVertex' z (overlay x y) == 'hasVertex' z x || 'hasVertex' z y
+-- 'vertexCount' (overlay x y) >= 'vertexCount' x
+-- 'vertexCount' (overlay x y) <= 'vertexCount' x + 'vertexCount' y
+-- 'edgeCount'   (overlay x y) >= 'edgeCount' x
+-- 'edgeCount'   (overlay x y) <= 'edgeCount' x   + 'edgeCount' y
+-- 'size'        (overlay x y) == 'size' x        + 'size' y
+-- 'vertexCount' (overlay 1 2) == 2
+-- 'edgeCount'   (overlay 1 2) == 0
+-- @
+overlay :: Graph a -> Graph a -> Graph a
+overlay (UG g1) (UG g2) = coerce (G.Overlay g1 g2)
+{-# INLINE overlay #-}
+
+-- | /Connect/ two graphs. This is a commutative and
+-- associative operation with the identity 'empty', which distributes over
+-- 'overlay' and obeys the decomposition axiom.
+-- Complexity: /O(1)/ time and memory, /O(s1 + s2)/ size. Note that the number
+-- of edges in the resulting graph is quadratic with respect to the number of
+-- vertices of the arguments: /m = O(m1 + m2 + n1 * n2)/.
+--
+-- @
+-- 'connect' x y               == 'connect' y x
+-- 'isEmpty'     (connect x y) == 'isEmpty'   x   && 'isEmpty'   y
+-- 'hasVertex' z (connect x y) == 'hasVertex' z x || 'hasVertex' z y
+-- 'vertexCount' (connect x y) >= 'vertexCount' x
+-- 'vertexCount' (connect x y) <= 'vertexCount' x + 'vertexCount' y
+-- 'edgeCount'   (connect x y) >= 'edgeCount' x
+-- 'edgeCount'   (connect x y) >= 'edgeCount' y
+-- 'edgeCount'   (connect x y) >= 'vertexCount' x * 'vertexCount' y
+-- 'edgeCount'   (connect x y) >= 'vertexCount' x * 'vertexCount' y `div` 2
+-- 'size'        (connect x y) == 'size' x        + 'size' y
+-- 'vertexCount' (connect 1 2) == 2
+-- 'edgeCount'   (connect 1 2) == 1
+-- @
+connect :: Graph a -> Graph a -> Graph a
+connect (UG g1) (UG g2) = coerce (G.Connect g1 g2)
+{-# INLINE connect #-}
+
+-- | Construct the graph comprising a given list of isolated vertices.
+-- Complexity: /O(L)/ time, memory and size, where /L/ is the length of the
+-- given list.
+--
+-- @
+-- vertices []            == 'empty'
+-- vertices [x]           == 'vertex' x
+-- 'hasVertex' x . vertices == 'elem' x
+-- 'vertexCount' . vertices == 'length' . 'Data.List.nub'
+-- 'vertexSet'   . vertices == Set . 'Set.fromList'
+-- @
+vertices :: [a] -> Graph a
+vertices = overlays . map vertex
+{-# INLINE vertices #-}
+
+-- TODO: Use a faster nubBy implementation with 'Data.Set'
+-- | Construct the graph from a list of edges.
+-- Complexity: /O(L^2)/ time, /O(L)/ memory and size, where /L/ is the length of the
+-- given list.
+--
+-- @
+-- edges []          == 'empty'
+-- edges [(x,y)]     == 'edge' x y
+-- edges [(x,y), (y,x)] == 'edge' x y
+-- @
+edges :: Eq a => [(a, a)] -> Graph a
```

Comment:
I don't think you need to do anything special here, just use `coerce1 G.edges`.

Reply:
This whole module is mostly just one big `coerce` :)

You'll only need to be careful about deconstructing the graph -- you'll need to treat `Connect` as a commutative operation.

Reply:
I see that now! I'm going to change everywhere I can!

Reply:
I'm following the approach where `Connect a b` means that 'a' is connected to 'b' and 'b' is connected to 'a'. I could write `connect` like this:
```Haskell
connect :: Graph a -> Graph a -> Graph a
connect = coerce2 G.connect . coerce2 (flip G.connect)
```
But I don't think this would be a good idea! But with this approach I would need to filter the duplicate edges on the ' edges' function. What would you prefer?

the `edges`function would be like this:

```Haskell
edges :: Eq a => [(a, a)] -> Graph a
edges = coerce1 G.edges . nubBy (\x y -> x == swap y)
```

`nubBy` can be optimized!

Reply:
> But with this approach I would need to filter the duplicate edges on the `edges` function.

Why? I don't understand. You don't need to filter out anything.

Reply:
`edges` in the Graph module is defined like this `edges = overlays . map (uncurry edge)` where `edge` is just `connect (vertex x) (vertex y)` if, for the `Undirected` module I use the same implementation of `edges` then for `edges [(1,2),(2,1)] == Overlay (Connect (Vertex 1) (Vertex 2)) (Connect (Vertex 2) (Vertex 1))`, right?

If we don't want to represent the redundant edges in the Graph structure itself I think we need to filter the redundant edges (that's what I meant with duplicate)

Reply:
> If we don't want to represent the redundant edges in the Graph structure itself

This was never our goal. It would be very hard to guarantee that we never have redundant vertices or edges in an expression. If you look at `Algebra.Graph`, for example, users can also call `edges` with repeated edges, but we never filter them out -- it would complicate the implementation.

Reply:
You are right it would be very hard! I will change my mindset! I was doing this thinking that we would never have redundant edges and that's not the case! Thanks for snapping me out of it

Reply:
With this in mind I think it will be a little hard to keep the API exactly the same as the `SymmetricRelation` one.. for example: `edges [(x,y), (y,x)] == 'edge' x y` this property does not hold.

Should I keep the semantics of the undirected graph and relax some of the properties that hold with `SymmetricRelation`?

Reply:
> EDIT: Well it actually holds because the Show,Eq, etc.. instances convert it to a symmetric relation.

Exactly. We never expose the internal implementation, which allows us to have some redundant parts in expressions.

Reply:
>  Should this property hold: edgeCount (edges [(1,2),(2,1)]) == 1 ?

Yes, of course, there is only one undirected edge in this graph.

Reply:
Okay! Thanks for your help!